res/controllers/Denon-Prime-4-scripts.js

```diff
@@ -0,0 +1,806 @@
+        },
+        shift: function() {
+            this.inKey = "play_stutter";
+            this.type = components.Button.prototype.types.push;
+        }
+    });
+
+    // Performance Pads
+    this.padGrid = new Prime4.PadSection(this, midiChannel - 4);
+
+    // Tempo Fader
+    this.tempoFader = new components.Pot({
+        midi: [0xB0 + midiChannel, 0x1F],
+        inKey: "rate",
+        invert: true,
+    });
+
+    // LED indicator when pitch fader is at centre.
+    this.tempoFeedback = new components.Button({
+        midi: [0x90 + midiChannel, 0x34],
+        outKey: "rate",
+        on: 0x02,
+        off: 0x00,
+        outValueScale: function(value) {
+            return value === 0 ? this.on : this.off;
+        },
+    });
+
+    // Keylock Button
+    this.keylockButton = new components.Button({
+        midi: [0x90 + midiChannel, 0x22],
+        key: "keylock",
+        type: components.Button.prototype.types.toggle,
+    });
+
+    /*
+    this.vinylButton = new components.Button({
+        midi: [0x90 + midiChannel, 0x23],
+        //TODO: Jog Wheel Functionality First
+        type: components.Button.prototype.types.toggle,
+    });
+    */
+
+    // Jog Wheel
+    this.jogWheel = new components.JogWheelBasic({
+        deck: script.deckFromGroup(this.currentDeck),
+        wheelResolution: 1000,
+        alpha: 1/8,
+        beta: 1/8/32,
+        rpm: 33 + 1/3,
+        // Instead of relative movements between this and the last position,
+        // the controller reports the absolute position of the wheel with
+        // 14-bit precision. Because of that, we need to reconstruct the value
+        // and then transform it into the relative directions expected by Mixxx.
+        inputWheelMSB: function(_channel, _control, value, _status, _group) {
+            this.wheelMSB = value;
+        },
+        inputWheelLSB: function(channel, control, value, status, group) {
+            this.inputWheel(channel, control, (this.wheelMSB << 7) + value, status, group);
+        },
+        previousPosition: null,
+        wrappingValue: Math.pow(2, 14),
+        inValueScale: function(value) {
+            // The first value of the controller will probably be random
+            // and thus we just have to swallow it until we have the second value
+            // to find the difference
+            if (this.previousPosition === null) {
+                this.previousPosition = value;
+                return 0;
+            }
+            // This finds the shortest distance between the current value
+            // and the last one, and preserves the orientation
+            const delta = value - this.previousPosition;
+            let remainder = ((delta % this.wrappingValue) + this.wrappingValue) % this.wrappingValue;
+            if (remainder * 2 > this.wrappingValue) {
+                remainder -= this.wrappingValue;
+            }
+            this.previousPosition = value;
+            return remainder;
+        },
```

Comment:
this should help with the problem where the jog is too sensitive.
```suggestion
        relativeFromAbsolute: function(value) {
            // The first value of the controller will probably be random
            // and thus we just have to swallow it until we have the second value
            // to find the difference
            if (this.previousPosition === null) {
                this.previousPosition = value;
                return 0;
            }
            // This finds the shortest distance between the current value
            // and the last one, and preserves the orientation
            const delta = value - this.previousPosition;
            let remainder = ((delta % this.wrappingValue) + this.wrappingValue) % this.wrappingValue;
            if (remainder * 2 > this.wrappingValue) {
                remainder -= this.wrappingValue;
            }
            this.previousPosition = value;
            return remainder;
        },
        jogScale: function (val) {
            // tweak the factor to your liking. It would probably sense to make it
            // user configurable via a variable at the top of the script.
            return val * 0.5;
        },
        inputWheel: function(channel, control, value, status, group) {
            value = this.relativeFromAbsolute(value);
            if (engine.isScratching(this.deck)) {
                engine.scratchTick(this.deck, value);
            } else {
                this.inSetValue(this.jogScale(value));
            }
        },
```